### ISSUE TITLE
FR: Compile the docs into the `jj` binary

### ISSUE THREAD

**Issue by ilyagr** [jj-vcs/jj#1294]
I think it would be nice for `jj` to be self-documenting. A relatively simple approach would be to create a `jj doc` command so that `jj doc tutorial` prints `docs/tutorial.md` via a pager, using the `include_str!` macro. `jj doc` or `jj doc list` would list the available topics. I'm hoping that `clap` will provide tab completion (though this is optional, especially for the first version).

For example, I expect `jj doc revsets` to be useful.

Alternatively, this could be part of the `jj help` command, but I couldn't figure out an easy way to do that with `clap`. Also, the output of `jj help` is already quite long.

This will also match the version of the documentation people see with the version of `jj` they use.

**Comment by martinvonz** [comment]
> I'd definitely start with making `jj help` more useful and self-contained. At least reference relevant docs from `help`.

Patches for that would be appreciated. I don't notice it much myself.

> 
> Maybe could have some kind of `jj help --topic FOO` command (along the lines of hg's `hg help FOO` vs `hg help --command FOO`)?

There actually used to be a command for that. I decided that it was less maintenance work to keep it in separate markdown files. There's probably a way of generating something that's readable in terminal from the markdown files. I think I looked into that a little bit a year or so.

**Comment by ilyagr** [comment]
> Maybe could have some kind of `jj help --topic FOO` command

I am a little apprehensive about touching `jj help` since it's auto-generated by `clap` and I don't want to replace it entirely.

> There's probably a way of generating something that's readable in terminal from the markdown files

At least at first, I was thinking of just sending them to the pager. There are also plenty of tools that try to render markdown more nicely in the terminal, we could later try finding a library that can be used from rust.

**Comment by gasche** [comment]
I ended up on this issue looking for `jj help revsets`, and then I learned that documentation on revsets is now available through `jj help --keyword revsets` as implemented by #4630. Keywords are not automatically populated from the `docs/` content, each of them is added manually (last addition in 8907e692d4de7c7caded771c20f9fe28fbcd3bbb). At the time of writing (jj 0.26.0), the following are supported:

1. bookmarks
2. config
3. filesets
4. glossary
5. revsets
6. templates
7. tutorial

If there is any other `docs/foo.md` file that you would like to see added, I suppose that sending a PR would be enough.

I suspect that the present issue could be closed.

**Comment by yuja** [comment]
> I suspect that the present issue could be closed.

Agreed. Thanks for the summary.

**Comment by gasche** [comment]
Note: after thinking about this yesterday, my personal conclusion is that I would personally prefer `jj docs revsets` over `jj help -k revsets`, in particular as it would make it much easier to advertise for this capability in the clap-generated documentation. The way [command/help.rs](https://github.com/jj-vcs/jj/blob/64ea8bee4767b59c85bb6daa830e4f11381bcd36/cli/src/commands/help.rs) works today is (very clever and) a bit baroque, and the fact that you need to run `jj help --help` rather than `jj help` to see a list of docs/keywords makes it harder for people to discover the feature.

@yuja, @Grillo-0  is the choice to go for `jj help -k revsets` final? Would it be possible to also offer `jj docs revsets` if you receive a PR for this? (I think it shouldn't be too hard to take bits of this one and reuse the existing logic in help.rs) It would be fine to have both commands coexist, for a while or indefinitely.

**Comment by yuja** [comment]
I personally don't think `jj docs` is more discoverable than `jj help -k`. It might be easier to implement, and `jj help docs` might sound a bit more natural than `jj help help`, but that shouldn't be a strong reason to deprecate `jj help -k` in favor of `jj docs`.

To improve discoverability, it's probably better to add hint to revset error message, mention `jj help -k revsets` in `jj help log`, etc.

**Comment by gasche** [comment]
For example, `jj --help` or `jj help` (they behave the same) does not mention the `-k | --keywords` parameter at all currently. If we had a `docs` subcommand, it would be listed there (without having to play tricks with clap, and in a good-visibility position), which is arguably the main entry point for embedded documentation/manpage/help about the tool.

**Comment by Grillo-0** [comment]
To meet in the middle, maybe we could just make `docs` an alias to `jj help -k`. It does not deprecate `jj help -k`, and could give a more natural way of finding what information

**Comment by yuja** [comment]
If people prefer `jj docs` (or `jj doc`?), I would suggest removing `jj help -k`. I don't think we need two commands that provide the same functionality.

(To be clear, I don't have any strong feeling. I personally find `jj help <KEYWORD>` is more discoverable, though.)

**Comment by yuja** [comment]
These problems are already solved, right? I thought your proposal was due to purely command discoverability. (I agree that `jj docs` would be easier to implement, btw.)

IIRC, the first version of `jj help` had a list of topics (or keywords), but we removed it because it seemed weird that the top-level help contained items that can't be used globally. So the list is now displayed in `jj help --help`. I assume that will be the same in `jj docs`. (`jj docs --help` will print a list of topics.)

`jj help <KEYWORD>` (without `-k`) can be supported if needed. We decided to add explicit flag `-k=KEYWORD` (instead of optional `--keyword`/`--command` for disambiguation), but I don't remember the reason.

**Comment by gasche** [comment]
I don't consider (2) to be fully solved right now; the blurb mentioning `jj help --help` is at the very end of the help output, after the part that users have trained themselves to skip, and `jj help --help` is very hard to discover otherwise -- it is a bizarre command. Compare to the output we would get with a proper subcommand:

```
$ jj -h
Jujutsu (An experimental VCS)

Usage: jj [OPTIONS] <COMMAND>

Commands:
  abandon           Abandon a revision
  [...]
  diffedit          Touch up the content changes in a revision with a diff editor
  docs              Show documentation on `jj` concepts, as found on the website.
  duplicate         Create new changes with the same content as existing ones
  [...]
  workspace         Commands for working with workspaces

Options:
  -h, --help     Print help (see more with '--help')
  -V, --version  Print version

Global Options:
  [...]
```

From there it's natural that `jj docs (-h|--help)` will show a listing of available documentation.

**Comment by Grillo-0** [comment]
> I don't consider (2) to be fully solved right now; the blurb mentioning `jj help --help` is at the very end of the help output, after the part that users have trained themselves to skip, and `jj help --help` is very hard to discover otherwise -- it is a bizarre command. Compare to the output we would get with a proper subcommand:
> 
> ```
> $ jj -h
> Jujutsu (An experimental VCS)
> 
> Usage: jj [OPTIONS] <COMMAND>
> 
> Commands:
>   abandon           Abandon a revision
>   [...]
>   diffedit          Touch up the content changes in a revision with a diff editor
>   docs              Show documentation on `jj` concepts, as found on the website.
>   duplicate         Create new changes with the same content as existing ones
>   [...]
>   workspace         Commands for working with workspaces
> 
> Options:
>   -h, --help     Print help (see more with '--help')
>   -V, --version  Print version
> 
> Global Options:
>   [...]
> ```
> 
> From there it's natural that `jj docs (-h|--help)` will show a listing of available documentation.

I agree that having a command entry would make it more discoverable. People will find it when looking for others subcommands.

I also think that, independent if we stay with `jj help -k` or switch to `jj docs`, it is a great idea to keep promoting the subcommand on others subcommands outputs, like `jj log` does. Personally, I think this is the most effective way of increasing discoverability, giving the tool at the moment the user needs.
